src/lib/chawathe98_matcher.rs

```diff
@@ -46,35 +46,2262 @@
+    pub cost_null: usize,
+    /// Check the cost of edge update which is Ok
+    pub cost_ok: usize,
+}
+
+/// implemenation for edge cost
+
+impl CostEdge {
+    /// Create a new cost edge
+    /// This shows the cost for each edge defined by the user
+    pub fn new(
+        cost_update_1: usize,
+        cost_insert_1: usize,
+        cost_delete_1: usize,
+        cost_move_1: usize,
+        cost_copy_1: usize,
+        cost_glue_1: usize,
+        cost_null_1: usize,
+        cost_ok_1: usize,
+    ) -> CostEdge {
+        CostEdge {
+            cost_update: cost_update_1,
+            cost_insert: cost_insert_1,
+            cost_delete: cost_delete_1,
+            cost_move: cost_move_1,
+            cost_copy: cost_copy_1,
+            cost_glue: cost_glue_1,
+            cost_null: cost_null_1,
+            cost_ok: cost_ok_1,
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+/// Edge construct which shows the edge between two nodes and their value
+pub struct Edge {
+    /// from node shows the node id from the From-Tree
+    pub from_node: NodeId<FromNodeId>,
+    /// to node shows the node id from To-Tree
+    pub to_node: NodeId<ToNodeId>,
+    /// The value of the edge
+    pub value: usize,
+    /// The edge type
+    pub edge_type: EdgeType,
+}
+
+/// Implementation for Edges
+impl Edge {
+    /// Create a new edge
+    pub fn new(
+        new_from_node: NodeId<FromNodeId>,
+        new_to_node: NodeId<ToNodeId>,
+        new_value: usize,
+        new_edge_type: EdgeType,
+    ) -> Edge {
+        Edge {
+            from_node: new_from_node,
+            to_node: new_to_node,
+            value: new_value,
+            edge_type: new_edge_type,
+        }
+    }
+}
+
+/// A store of mappings between nodes in different arenas.
+/// Direction is important.
+#[derive(Debug, Clone)]
+pub struct MappingStore2<T: Clone + Debug> {
+    ///List Of edges
+    pub list_edges: RefCell<Vec<Edge>>,
+    /// Source arena (treat as immutable).
+    pub from_arena: RefCell<Arena<T, FromNodeId>>,
+    /// Destination arena (treat as immutable).
+    pub to_arena: RefCell<Arena<T, ToNodeId>>,
+    /// List of nodes to be inserted
+    pub list_node_inserted: RefCell<Vec<NodeId<ToNodeId>>>,
+    /// List of nodes to be deleted
+    pub list_node_deleted: RefCell<Vec<NodeId<FromNodeId>>>,
+    /// Find the Cost of all edges
+    pub all_edge_cost: RefCell<CostEdge>,
+}
+
+impl<T: Clone + Debug + Eq + 'static> MappingStore2<T> {
+    /// Implementation of new in mapping store the second version
+    pub fn new(base: Arena<T, FromNodeId>, diff: Arena<T, ToNodeId>) -> MappingStore2<T> {
+        MappingStore2 {
+            from_arena: RefCell::new(base),
+            to_arena: RefCell::new(diff),
+            list_edges: RefCell::new(Vec::new()),
+            list_node_inserted: RefCell::new(Vec::new()),
+            list_node_deleted: RefCell::new(Vec::new()),
+            all_edge_cost: RefCell::new(CostEdge::new(1, 1, 1, 1, 1, 1, 1, 1)),
```

Comment:
This is fine, but because it's a bit of a temporary hack, let's put a comment in to say "Currently, we set all the edge costs to be 1, but in future they will be configurable."